File: apps/ui/src/Types/modules/git-branches.ts
/**
 * The branch half of the git API.
 *
 * A fourth module beside `api.ts`, `git.ts` and `git-history.ts`, mirroring
 * the fourth Rust file: `GitBranchTransport` is a supertrait of
 * `GitTransport`, and `GitBranchClient` is extended by `GitClient`. One
 * surface for callers, one readable file per surface.
 *
 * Re-exported through `@/Types`, so nothing imports from here directly.
 */
import type { CreateBranchRequest, GitBranch } from "../generated";

/** Tauri command names. The only place these strings are written down. */
export const GIT_BRANCH_COMMANDS = {
  branches: "git_branches",
  checkout: "git_checkout",
  createBranch: "git_create_branch",
  deleteBranch: "git_delete_branch",
} as const;

export type GitBranchCommand =
  (typeof GIT_BRANCH_COMMANDS)[keyof typeof GIT_BRANCH_COMMANDS];

/** Command argument payloads, one per command that takes arguments. */
export type GitBranchNameArgs = { name: string };
export type GitCreateBranchArgs = { request: CreateBranchRequest };
export type GitDeleteBranchArgs = GitBranchNameArgs & { force: boolean };

/**
 * Mirrors `mino_core::transport::GitBranchTransport`.
 *
 * What separates these from everything before them: `checkout` **changes files
 * under the other three panes**. Nothing on this interface refreshes anything -
 * the caller fires one "git state changed" event afterwards and every pane
 * re-reads at the same moment, rather than each guessing. See
 * `features/git/context/GitRefreshContext`.
 */
export interface GitBranchClient {
  /**
   * Every branch the picker can offer, local and remote in one call, each with
   * its upstream, ahead/behind counts and tip commit.
   *
   * A repository with no commits answers with an empty list rather than
   * rejecting: `git init` and nothing since is a state, not a failure.
   */
  branches(): Promise<GitBranch[]>;

  /**
   * Switch HEAD, and the working tree with it.
   *
   * Faithful to `git checkout <name>`, which means a *remote-tracking* name
   * detaches HEAD rather than creating a local branch. Callers that mean
   * "start working on this remote branch" send its short name and let git's
   * own DWIM create the tracking branch - see `useBranches`.
   *
   * Rejects, changing nothing, when the branch is not there or the working
   * tree would be overwritten - two different sentences, because they are two
   * different things the reader can act on. Git never switches halfway, so a
   * rejection means the repository is exactly as it was.
   *
   * Warning about an unsaved editor draft happens **before** this is called.
   * Git knows nothing about a buffer that was never written, so nothing below
   * this line can protect one.
   */
  checkout(name: string): Promise<void>;

  /**
   * Create a branch, and switch to it when `request.checkout` is set.
   *
   * Resolves with the branch that now exists rather than `void`, so the picker
   * shows what git made instead of assuming its own request came true. A
   * duplicate name rejects.
   */
  createBranch(request: CreateBranchRequest): Promise<GitBranch>;

  /**
   * Delete a branch. Without `force` git refuses one whose commits are nowhere
   * else, and refuses the branch you are on.
   *
   * With `force` this **destroys commits**, so callers confirm first and say
   * what will be lost - the discard rule, applied to a branch.
   */
  deleteBranch(name: string, force: boolean): Promise<void>;
}

File: apps/desktop/src-tauri/src/commands/git_branches.rs
//! The branch commands. Dispatch only, like every command in this module.
//!
//! Split from `git.rs` the way the core trait is split: branches are the first
//! thing on this interface that changes files under the *other* panes, and
//! keeping them in their own file is what makes that visible rather than
//! buried in a list.
//!
//! Nothing here decides when the panes refresh. That is one event in the UI,
//! fired after the call returns, so every pane learns at the same moment
//! instead of each guessing - see `docs/mino-workbench/git-module.md`.

use mino_core::types::{CreateBranchRequest, GitBranch};
use mino_core::TransportError;
use tauri::State;

use crate::state::AppState;

use super::git::git;

#[tauri::command]
pub async fn git_branches(state: State<'_, AppState>) -> Result<Vec<GitBranch>, TransportError> {
    let transport = state.current()?;
    git(&transport)?.branches().await
}

/// Switches HEAD. The warning about an unsaved draft is the UI's job and has
/// already happened by the time a call reaches here - git knows nothing about
/// a buffer that was never written.
#[tauri::command]
pub async fn git_checkout(state: State<'_, AppState>, name: String) -> Result<(), TransportError> {
    let transport = state.current()?;
    git(&transport)?.checkout(&name).await
}

#[tauri::command]
pub async fn git_create_branch(
    state: State<'_, AppState>,
    request: CreateBranchRequest,
) -> Result<GitBranch, TransportError> {
    let transport = state.current()?;
    git(&transport)?.create_branch(request).await
}

/// `force` is the destructive form: it deletes a branch whose commits are
/// nowhere else. The confirmation is the UI's, the same way discard's is.
#[tauri::command]
pub async fn git_delete_branch(
    state: State<'_, AppState>,
    name: String,
    force: bool,
) -> Result<(), TransportError> {
    let transport = state.current()?;
    git(&transport)?.delete_branch(&name, force).await
}
